# Section 4: Working with DataFrames
Dataframes are the blockbuster data struction in Pandas. Many of the methods and techniques we learned for Series are applicable to Dataframes, but they become even more powerful!
* We'll learn how to clean data and prepare it for analysis
* We'll also learn several Dataframe manipulation techniques

In [1]:
import pandas as pd
import numpy as np

## What is a DataFrame?
A **Dataframe** is quite simply a table of data that contains a collection of rows and columns. They are generally two-dimensional (as opposed to Series which are one-dimensional), with labeled indices and columns. However, by using multiple indices, they are also able to accommodate multi-dimensional data. More on that later.

What this practically means is that we need to specify more than one piece of information in order to identify a specific datapoint in the Dataframe. This contrasts with Series, in which only one piece of information is needed (usually the index label or index position).

The `.ndim` attribute gives us the number of dimensions for a Series or Dataframe. Series have 1 dimension and Dataframes have 2 dimensions
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.ndim.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.ndim.html

Recall the `.shape` attribute from Dataframe, which returns a tuple telling us the dimensionality of the dataframe, that is, the number of rows and columns.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html

Both Series and Dataframes are a collection of values with associated labels. But whereas Series have labeled indices only, Dataframes have labeled indices and columns

Each column in a Dataframe is actually Series. That is, the Series object is the data structure that comprises the column of a Dataframe.

Unlike Series, Dataframes can be *heterogenous*. That is, each column of a DataFrame can be a completely different data type. 
* DataFrames themselves DO NOT have a datatype, since they are a collection of different Series. Thus, the `dtype` attribute on a Dataframe will return an attribute error.
* Instead, the `dtypes` attribute may be used on a Dataframe, which will return a Series with the names of the columns as the index labels and the data type as the values.
 * https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html

## Creating a Dataframe

Just like Series, Dataframes can be created manually. The typical way to do this is to construct a list of each of your columns, and then combine them into a Dataframe.

In [2]:
names = ['Olga','Andrew','Brian','Telulah','Nicole','Tilda']
age = [29, 21, 45, 23, 39, 46]
married = [False, True, True, True, False, True]

After creating the lists, we can build a Dataframe using the `pd.DataFrame()` function. We pass in a dictionary whose keys are the names of the columns, and the values are the names of the lists we created.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html

In [3]:
pd.DataFrame({'name': names, 'age': age, 'married': married})

name  age  married
0     Olga   29    False
1   Andrew   21     True
2    Brian   45     True
3  Telulah   23     True
4   Nicole   39    False
5    Tilda   46     True

Keep in mind that the lists that you construct your dataframe out of MUST be of equal length. If they are not, you will encounter a `ValueError`


In [4]:
# names2 = ['Olga','Andrew','Brian','Telulah','Nicole','Tilda', 'Ryan']
# pd.DataFrame({'name': names2, 'age': age, 'married': married})

## BONUS - Four More Ways to Build DataFrames

There are other ways for creating DataFrames as well. The first is by building from a **dictionary of tuples** (instead of a dictionary of lists)

In [5]:
tuple_names = tuple(names)
tuple_ages = tuple(age)
tuple_married = tuple(married)

In [6]:
pd.DataFrame({'name': tuple_names, 'age': tuple_ages, 'married': tuple_married})

name  age  married
0     Olga   29    False
1   Andrew   21     True
2    Brian   45     True
3  Telulah   23     True
4   Nicole   39    False
5    Tilda   46     True

The second of the alternative methods is to pass in a **dictionary of Pandas Series**. All we really need to do is build a set of Series, and combine them using the `pd.DataFrame` method

In [7]:
series_names = pd.Series(data = names)
series_ages = pd.Series(data = age)
series_married = pd.Series(data = married)

In [8]:
pd.DataFrame({'name': series_names, 'age': series_ages, 'married': series_married})

name  age  married
0     Olga   29    False
1   Andrew   21     True
2    Brian   45     True
3  Telulah   23     True
4   Nicole   39    False
5    Tilda   46     True

Slightly more challenging is to use a **dictionary of dictionaries**. The strategy:
1. Replicate the upper left section of the dataframe in order to understand what shape of data you are targeting
2. Then, extend that same structure to the rest of the data programmatically so that we don't have to type it all out

In [9]:
pd.DataFrame({'names' : {0: 'Olga', 1: 'Andrew'}})

names
0    Olga
1  Andrew

From this starter, we use the built-in Python function `enumerate()` to traverse the list and access each item in turn.

In [10]:
list(enumerate(names))

[(0, 'Olga'),
 (1, 'Andrew'),
 (2, 'Brian'),
 (3, 'Telulah'),
 (4, 'Nicole'),
 (5, 'Tilda')]

We'll use dictionary comprehension to create key:value pairs for each of these enumerated names.

In [11]:
dict_names = {key:value for key, value in enumerate(names)}
dict_names

{0: 'Olga', 1: 'Andrew', 2: 'Brian', 3: 'Telulah', 4: 'Nicole', 5: 'Tilda'}

Let's take care of the rest of the columns, converting them all to dictionaries of keys and values

In [12]:
dict_ages = {key:value for key, value in enumerate(age)}
print(dict_ages)
dict_married = {key:value for key, value in enumerate(married)}
print(dict_married)

{0: 29, 1: 21, 2: 45, 3: 23, 4: 39, 5: 46}
{0: False, 1: True, 2: True, 3: True, 4: False, 5: True}


Note that we could have done this using a function, which helps streamline things a big. If you needed to change how the enumeration is carried out, you need only modify your one function instead of modifying each conversion individually:

In [13]:
def convert_list_to_dict(l):
  return {key:value for key, value in enumerate(l)}

Now all we have to do is construct the DataFrame!

In [14]:
pd.DataFrame({'name': dict_names,
              'ages': dict_ages,
              'married': dict_married})

name  ages  married
0     Olga    29    False
1   Andrew    21     True
2    Brian    45     True
3  Telulah    23     True
4   Nicole    39    False
5    Tilda    46     True

Finally, we can take a row-centric approach that builds a DataFrame row by row (as opposed to the column-centric approaches above that build DataFrames column-by-column) by using a **list of dictionaries**

Let's start by replicating a single row:

In [15]:
pd.DataFrame([{'name':'Olga', 'age':29, 'married': False}])

name  age  married
0  Olga   29    False

Next we use the `zip()` method to combine multiple iterables (in this case `names`, `age`, and `married`, into a list of tuples

In [16]:
list(zip(names, age, married))

[('Olga', 29, False),
 ('Andrew', 21, True),
 ('Brian', 45, True),
 ('Telulah', 23, True),
 ('Nicole', 39, False),
 ('Tilda', 46, True)]

Now we have to build a list of dictionaries. Each row will be its own dictionary, and we can do this using `zip()` and list comprehension

In [17]:
rowwise = [{'name': name, 'age': age, 'married':married} for name, age, married in zip(names, age, married)]
rowwise

[{'age': 29, 'married': False, 'name': 'Olga'},
 {'age': 21, 'married': True, 'name': 'Andrew'},
 {'age': 45, 'married': True, 'name': 'Brian'},
 {'age': 23, 'married': True, 'name': 'Telulah'},
 {'age': 39, 'married': False, 'name': 'Nicole'},
 {'age': 46, 'married': True, 'name': 'Tilda'}]

Finally, we combine them into a DataFrame!

In [18]:
pd.DataFrame(rowwise)

name  age  married
0     Olga   29    False
1   Andrew   21     True
2    Brian   45     True
3  Telulah   23     True
4   Nicole   39    False
5    Tilda   46     True

There are additional methods for building DataFrames, but we won't be covering them here. Again, in practice you will primarily be reading data into Pandas instead of building it from within.

## The `info()` method


The `info()` method gives you a brief summary of what a DataFrame contains. it is one of the first things you should do when getting ready to work with a new DataFrame.

`info()` prints information including the index dtype, the column names counts of non-null values, and memory usage.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html

In [19]:
df = pd.DataFrame({'name': names, 'age': age, 'married': married})
df

name  age  married
0     Olga   29    False
1   Andrew   21     True
2    Brian   45     True
3  Telulah   23     True
4   Nicole   39    False
5    Tilda   46     True

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     6 non-null      object
 1   age      6 non-null      int64 
 2   married  6 non-null      bool  
dtypes: bool(1), int64(1), object(1)
memory usage: 230.0+ bytes


The above example was a small DataFrame, but for large DataFrames even the `info()` summary may get very long and overwhelming. We can set the `verbose` parameter to *False* and  the individual column names will not be printed. Additionally, the different dtypes will be summarized instead of listed for each column.

In [21]:
df.info(verbose = False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Columns: 3 entries, name to married
dtypes: bool(1), int64(1), object(1)
memory usage: 230.0+ bytes


We can also set the `maxcols` parameter to limit the number of columns that will be described by the verbose setting. If the number of columns in the DataFrame exceeds this parameter, it will convert to a non-verbose summary.

In [22]:
print(df.info(max_cols=4))
print('')
print(df.info(max_cols=2))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     6 non-null      object
 1   age      6 non-null      int64 
 2   married  6 non-null      bool  
dtypes: bool(1), int64(1), object(1)
memory usage: 230.0+ bytes
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Columns: 3 entries, name to married
dtypes: bool(1), int64(1), object(1)
memory usage: 230.0+ bytes
None


Perhaps the most useful parameter from `info()` is `memory_usage`, where it will estimate the DataFrame's size in memory. If we set `memory_usage` to *deep*, the method will in calculate real memory usage with depp introspection. Otherwise, teh estimation will be made based on column dtype and number of rows, asssuming values consume the same memory amount for corresponding dtypes.

Using *deep* `memory_usage` is slower and requires more computation.

In [23]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     6 non-null      object
 1   age      6 non-null      int64 
 2   married  6 non-null      bool  
dtypes: bool(1), int64(1), object(1)
memory usage: 557.0 bytes


## Reading in Nutritional Data

The remainder of this section will rely on a sizeable dataset for roughyl 9000 food items and their nutritional information.

dataurl = https://andybek.com/pandas-nutrition

Import the data

In [24]:
dataurl = 'https://andybek.com/pandas-nutrition'
nutrition = pd.read_csv(dataurl)
nutrition.head(10)

Unnamed: 0                name serving_size  ...  caffeine theobromine    water
0           0          Cornstarch        100 g  ...   0.00 mg     0.00 mg   8.32 g
1           1        Nuts, pecans        100 g  ...   0.00 mg     0.00 mg   3.52 g
2           2       Eggplant, raw        100 g  ...   0.00 mg     0.00 mg  92.30 g
3           3      Teff, uncooked        100 g  ...         0           0   8.82 g
4           4     Sherbet, orange        100 g  ...   0.00 mg     0.00 mg  66.10 g
5           5    Cauliflower, raw        100 g  ...   0.00 mg     0.00 mg  92.07 g
6           6    Taro leaves, raw        100 g  ...   0.00 mg     0.00 mg  85.66 g
7           7   Lamb, raw, ground        100 g  ...   0.00 mg     0.00 mg  59.47 g
8           8   Cheese, camembert        100 g  ...   0.00 mg     0.00 mg  51.80 g
9           9  Vegetarian fillets        100 g  ...   0.00 mg     0.00 mg  45.00 g

[10 rows x 77 columns]

This is a fairly large DataFrame. The info method shows that it has 77 columns and 8789 entries. It takes up about 39.2 MB of memory.

Note that one of the dtypes is called `object`. This has to do with the fact that for many of the columns,the units are embedded with the values. Therefore, Pandas cannot treat them as numericals. Instead, they are automatically converted into strings, and Pandas treats strings as `object` datatypes.

In [25]:
nutrition.info(verbose = False, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8789 entries, 0 to 8788
Columns: 77 entries, Unnamed: 0 to water
dtypes: int64(3), object(74)
memory usage: 39.2 MB


## Cleanup: Removing Duplicated index
Let's take our first crack at cleaning data! If you look at your DataFrame, there is an extra column called `Unnamed: 0`, which appears to be a repitition of the index labels. This is NOT an index column, but rather just another column.

In [26]:
nutrition['Unnamed: 0']

0          0
1          1
2          2
3          3
4          4
        ... 
8784    8784
8785    8785
8786    8786
8787    8787
8788    8788
Name: Unnamed: 0, Length: 8789, dtype: int64

We should remove this column as it is not useful. One option is to use the `drop()` method. All one must do is specify the name of the entity to be dropped as well as the `axis`. Remember that DataFrames are two-dimensional, and so the axis on which a property is to be dropped must be specified. After all, the name could refer to either an index label or a column name, and you have to tell Pandas which one it is.

By default, this will return a new DataFrame with the indicated column(s) dropped.

In [27]:
nutrition.drop('Unnamed: 0', axis = 1)

name  ...    water
0                                            Cornstarch  ...   8.32 g
1                                          Nuts, pecans  ...   3.52 g
2                                         Eggplant, raw  ...  92.30 g
3                                        Teff, uncooked  ...   8.82 g
4                                       Sherbet, orange  ...  66.10 g
...                                                 ...  ...      ...
8784  Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  72.51 g
8785  Lamb, cooked, separable lean only, composite o...  ...  59.95 g
8786  Lamb, raw, separable lean and fat, composite o...  ...  59.80 g
8787  Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  73.43 g
8788  Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  73.43 g

[8789 rows x 76 columns]

Another option is to designated the unwanted column as the index using the `set_index` method. That works for this example because the actual index and the unwanted column are the same. In practice, you will want to ensure that the column you set as the index is truely appropriate as the index.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_index.html

In any case, the column indicated will replace the integer index column.

In [28]:
nutrition.set_index('Unnamed: 0')

name  ...    water
Unnamed: 0                                                     ...         
0                                                  Cornstarch  ...   8.32 g
1                                                Nuts, pecans  ...   3.52 g
2                                               Eggplant, raw  ...  92.30 g
3                                              Teff, uncooked  ...   8.82 g
4                                             Sherbet, orange  ...  66.10 g
...                                                       ...  ...      ...
8784        Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  72.51 g
8785        Lamb, cooked, separable lean only, composite o...  ...  59.95 g
8786        Lamb, raw, separable lean and fat, composite o...  ...  59.80 g
8787        Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  73.43 g
8788        Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  73.43 g

[8789 rows x 76 columns]

A third approach, advocated by the instructor, is to recognize that the `Unnamed: 0` column was almost certainly an index for this dataset, and that we failed to notice that when reading in the data. If we do recognize it, we can let Pandas know in the `read_csv()` call that this column (or any column for that matter) is the index column by using the `index_col` parameter.

In this example, specifiying `index_col=0` tells Pandas that the first column (at index 0) is the index column. This column will be set as the index. If we do not specify this, then Pandas will create a default index column with integers.

In [29]:
nutrition = pd.read_csv(dataurl, index_col='Unnamed: 0')
nutrition.head(10)

name serving_size  calories  ... caffeine theobromine    water
0          Cornstarch        100 g       381  ...  0.00 mg     0.00 mg   8.32 g
1        Nuts, pecans        100 g       691  ...  0.00 mg     0.00 mg   3.52 g
2       Eggplant, raw        100 g        25  ...  0.00 mg     0.00 mg  92.30 g
3      Teff, uncooked        100 g       367  ...        0           0   8.82 g
4     Sherbet, orange        100 g       144  ...  0.00 mg     0.00 mg  66.10 g
5    Cauliflower, raw        100 g        25  ...  0.00 mg     0.00 mg  92.07 g
6    Taro leaves, raw        100 g        42  ...  0.00 mg     0.00 mg  85.66 g
7   Lamb, raw, ground        100 g       282  ...  0.00 mg     0.00 mg  59.47 g
8   Cheese, camembert        100 g       300  ...  0.00 mg     0.00 mg  51.80 g
9  Vegetarian fillets        100 g       290  ...  0.00 mg     0.00 mg  45.00 g

[10 rows x 76 columns]

## The `sample()` Method

`sample()` is called on a DataFrame and returns back a random record/observation from that DataFrame
* You can also use the `random_state` parameter to "choose" which record(s) to return. This provides a seed for a random number generator. For any given random state, it will return the same record(s) each time
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html

In [30]:
nutrition.sample()

name serving_size  ...  theobromine    water
1877  Mushrooms, stir-fried, shiitake        100 g  ...      0.00 mg  87.74 g

[1 rows x 76 columns]

In [31]:
nutrition.sample(random_state=2)

name  ...    water
7652  Beef, raw, all grades, trimmed to 0" fat, sepa...  ...  72.96 g

[1 rows x 76 columns]

You can use a comparison operator to prove that the same random state will return the same records every time.

In [32]:
nutrition.sample(random_state=2) == nutrition.sample(random_state=2)

name  serving_size  calories  ...  caffeine  theobromine  water
7652  True          True      True  ...      True         True   True

[1 rows x 76 columns]

You can choose how many records to return using the `n` parameter. It defaults to 1


In [33]:
nutrition.sample(n = 5)

name  ...    water
8294  Pork, braised, cooked, with added solution, se...  ...  56.13 g
5841  Veal, braised, cooked, thymus, variety meats a...  ...  73.63 g
1349                             Shallots, freeze-dried  ...   2.00 g
2637                Potatoes, prepared, frozen, o'brien  ...  61.96 g
8263  Lamb, raw, trimmed to 1/8" fat, separable lean...  ...  67.33 g

[5 rows x 76 columns]

You can also use the `frac` parameter to indicate what percentage of the data you want to randomly sample. It cannot be used at the same time as the `n` parameter. Think about why that makes sense.

In [34]:
nutrition.sample(frac = 0.01)

name  ...    water
480                                     Fish, raw, ling  ...  79.63 g
8679  Beef, roasted, cooked, select, trimmed to 1/8"...  ...  49.40 g
8739  Beef, raw, select, trimmed to 0" fat, separabl...  ...  73.01 g
522                                      Jackfruit, raw  ...  73.46 g
602                                        Potato flour  ...   6.52 g
...                                                 ...  ...      ...
1917                     Cocoa, unsweetened, dry powder  ...   3.00 g
6671  Artichokes, with salt, drained, boiled, cooked...  ...  84.08 g
7896  Beef, raw, select, trimmed to 0" fat, separabl...  ...  70.52 g
7011  Yogurt, fortified with vitamin D, 11 grams pro...  ...  79.00 g
3220          Snacks, nonfat, fruit-filled, granola bar  ...  13.70 g

[88 rows x 76 columns]

In [35]:
nutrition.sample(frac = 0.01).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88 entries, 5991 to 879
Data columns (total 76 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   name                         88 non-null     object
 1   serving_size                 88 non-null     object
 2   calories                     88 non-null     int64 
 3   total_fat                    88 non-null     object
 4   saturated_fat                64 non-null     object
 5   cholesterol                  88 non-null     object
 6   sodium                       88 non-null     object
 7   choline                      88 non-null     object
 8   folate                       88 non-null     object
 9   folic_acid                   88 non-null     object
 10  niacin                       88 non-null     object
 11  pantothenic_acid             88 non-null     object
 12  riboflavin                   88 non-null     object
 13  thiamin                      88 n

## BONUS - Sampling with Replacement or Weights

The sampling method can get even fancier.
* **Sampling with replacement** refers to the act of placing a record back into the population after it is selected, such that the probability of selecting that item in subsequent selections is unchanged.
* This makes it possible to pick the exact same record multiple times
* This is also known as **bootstrapping** in statistics


Within the `sample()` method, we can use the `replace` parameter to determine whether we sample with replacement. By setting `replace` to `True`, there is a chance to select the same record more than once.

In [36]:
nutrition.sample(n=3, replace = True)

name  ...    water
85                                SILK Original Creamer  ...  86.29 g
8345  Beef, grilled, cooked, all grades, trimmed to ...  ...  52.29 g
8108  Beef, raw, all grades, trimmed to 1/8" fat, se...  ...  73.32 g

[3 rows x 76 columns]

**Weighted sampling** refers to weighting our records for sampling. The higher the weight, the higher the likelihood of being selected.

Within the `sample()` method we do this using the `weights` parameter.
* If `weights` is not used, all reocrds will have equal probability weighting - that is, there will be equal likelihood of selecting any given record.
* This becomes useful when the indices have meaningful labels that would justify giving more weight to some records and less weight to others.

As an example, we start by creating a Series of weights, indexed by the label of the record that we want to weight in the main DataFrame

In [37]:
weights = pd.Series(data = [10, 10, 10, 1, 2], index=[7, 17, 29, 5, 6])
weights

7     10
17    10
29    10
5      1
6      2
dtype: int64

Now let's use these weights for our sampling. Notice that the higher weighted indices are more likely to be selected.

In [38]:
nutrition.sample(n=3, weights= weights)

name serving_size  calories  ... caffeine theobromine    water
6         Taro leaves, raw        100 g        42  ...  0.00 mg     0.00 mg  85.66 g
17  Peppers, raw, jalapeno        100 g        29  ...  0.00 mg     0.00 mg  91.69 g
5         Cauliflower, raw        100 g        25  ...  0.00 mg     0.00 mg  92.07 g

[3 rows x 76 columns]

## BONUS - How Are "Random Numbers" Generated?


How does Pandas, or any other RNGs for that matter, how does it decide what number to return to us? When we observe random phenomena in the real world, we are witness "natural attributes", or **true randomness**. That is, **what we observe now tells us nothing about what will come next**. You will get no predictiveness out of this.

In practice, it is difficult to sample a truly random attribute.

At random.org, they use atmospheric measurements and use that stream of randomness to generate numbers, streams, names, etc. It is true randomness that is supported by (as far as we can tell) a truly random natural phenomenon.

By contrast, computer programs create **pseudorandomness**. Computers are generally very algorithmic, following a set of rules for generating numbers. They cannot on their own generate numbers. For this reason, computer scientists have generated pseudo-random number generators (**PRNG**s). They start with a piece of input, and then from there they predictively generate a sequence of numbers. The catch is that these generators behave they eventually repeat themselves in a non-random fashion.

Python uses the Mersenne Twister, one of the most widely-used PRNGs.

Circling back to our `sample()` method in Pandas, by seeding a `random_state`, we lock down the starting input. Because the PRNG is deterministic, the number or number sequence that is generated with that particular seed state will always be the same.

Some useful links on random number generators:
* https://www.random.org/randomness/
* https://en.wikipedia.org/wiki/Mersenne_Twister
* https://www.random.org/

## DataFrame Axes

Recall that DataFrames are 2-dimensional. These two dimensions are typically the *index labels* and the *columnms*. Thus, the number of attributes that you would need to specify to select as specific observation is two - we need to provide a row label and a column label. Think of them as coordinates as sorts.

The `axes` attribute returns a list representing the axes of the dataframe. The list has two entries
1. A Pandas Index object representing the index labels
2. A Pandas Index object representing the column labels
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.axes.html

In [39]:
nutrition.axes

[Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
             ...
             8779, 8780, 8781, 8782, 8783, 8784, 8785, 8786, 8787, 8788],
            dtype='int64', length=8789),
 Index(['name', 'serving_size', 'calories', 'total_fat', 'saturated_fat',
        'cholesterol', 'sodium', 'choline', 'folate', 'folic_acid', 'niacin',
        'pantothenic_acid', 'riboflavin', 'thiamin', 'vitamin_a',
        'vitamin_a_rae', 'carotene_alpha', 'carotene_beta',
        'cryptoxanthin_beta', 'lutein_zeaxanthin', 'lucopene', 'vitamin_b12',
        'vitamin_b6', 'vitamin_c', 'vitamin_d', 'vitamin_e', 'tocopherol_alpha',
        'vitamin_k', 'calcium', 'copper', 'irom', 'magnesium', 'manganese',
        'phosphorous', 'potassium', 'selenium', 'zink', 'protein', 'alanine',
        'arginine', 'aspartic_acid', 'cystine', 'glutamic_acid', 'glycine',
        'histidine', 'hydroxyproline', 'isoleucine', 'leucine', 'lysine',
        'methionine', 'phenylalanine', 'proline', 'ser

Looking at them in isolation. We can, for example, get the third index label by accessing the first element of the list (the Index labels Pandas object) and then accessing the third item in that list as follows:

In [40]:
nutrition.axes[0][3]

3

A more direct way of doing this would be the access the index directly using the `index` attribute followed by bracket notation:

In [41]:
nutrition.index[3]

3

We can also, for example, the the 69th column label by accessing the second item in the `axes` return (the column labels Pandas object) and then accessing the 69th item in that list:

In [42]:
nutrition.axes[1][69]

'polyunsaturated_fatty_acids'

A more direct way to do this would be to access the columns directly using the `columns` attribute followed by bracket notation.

In [43]:
nutrition.columns[69]

'polyunsaturated_fatty_acids'

Generally speaking, the `axis` parameter shows us in MANY Pandas DataFrame methods. In any method that supports this parameter, `0` refers to **rows** and `1` refers to **columns**
* These methods also support `row` and `column` in place of `0` and `1`, though the instructor prefers the integer-based references

## Changing the Index of a DataFrame

By default, DataFrames have integer-based indices. The datatype is `Int64Index`. 



In [44]:
nutrition.index

Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
            ...
            8779, 8780, 8781, 8782, 8783, 8784, 8785, 8786, 8787, 8788],
           dtype='int64', length=8789)

`RangeIndex` which we saw earlier is a special case of the `Int64Index`. It is a more optimized version of `Int64Index` in which the entire sequence is defined by explicit start and stop indices and steps.

In [45]:
pd.RangeIndex(start=0, stop=8789, step=1)

RangeIndex(start=0, stop=8789, step=1)

Let's assign this as the index of our DataFrame. Notice the change in the datatype.

In [46]:
print(type(nutrition.index))
nutrition.index = pd.RangeIndex(start=0, stop=8789, step=1)
print(type(nutrition.index))

<class 'pandas.core.indexes.numeric.Int64Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


Now, what if we wanted to change the index from a series of integers to something else? One solution is to use the `set_index()` method on the DataFrame. We've seen this before.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_index.html

Here, we will use the name of the food as the index. Remember that `set_index()` returns a copy. To modify the existing DataFrame, set `inplace = True`

Also keep in mind that the `drop` parameter in `set_index` defaults to `True`, and so the column that is set as the new index is removed as a column in the DataFrame

In [47]:
nutrition.set_index('name')

serving_size  ...    water
name                                                             ...         
Cornstarch                                                100 g  ...   8.32 g
Nuts, pecans                                              100 g  ...   3.52 g
Eggplant, raw                                             100 g  ...  92.30 g
Teff, uncooked                                            100 g  ...   8.82 g
Sherbet, orange                                           100 g  ...  66.10 g
...                                                         ...  ...      ...
Beef, raw, all grades, trimmed to 0" fat, separ...        100 g  ...  72.51 g
Lamb, cooked, separable lean only, composite of...        100 g  ...  59.95 g
Lamb, raw, separable lean and fat, composite of...        100 g  ...  59.80 g
Beef, raw, all grades, trimmed to 0" fat, separ...        100 g  ...  73.43 g
Beef, raw, all grades, trimmed to 0" fat, separ...        100 g  ...  73.43 g

[8789 rows x 75 columns]

The `append` parameter in `set_index()` defaults to `False`. By setting it to `True`, the index will be appended in addition to the existing index, creating a multi-index DataFrame. We will revisit this concept later on.

In [48]:
multi_ind = nutrition.set_index('folic_acid', drop=False, append = True).set_index('name', drop=False, append = True).head(5)
multi_ind

name serving_size  ...  theobromine    water
  folic_acid name                                           ...                      
0 0.00 mcg   Cornstarch            Cornstarch        100 g  ...      0.00 mg   8.32 g
1 0.00 mcg   Nuts, pecans        Nuts, pecans        100 g  ...      0.00 mg   3.52 g
2 0.00 mcg   Eggplant, raw      Eggplant, raw        100 g  ...      0.00 mg  92.30 g
3 0          Teff, uncooked    Teff, uncooked        100 g  ...            0   8.82 g
4 0.00 mcg   Sherbet, orange  Sherbet, orange        100 g  ...      0.00 mg  66.10 g

[5 rows x 76 columns]

You can run the `axes` attribute on a multi-dimensional DataFrame and you will see that the index axis (axis 0) is now a `MultiIndex` datatype.

In [49]:
multi_ind.axes

[MultiIndex([(0, '0.00 mcg',      'Cornstarch'),
             (1, '0.00 mcg',    'Nuts, pecans'),
             (2, '0.00 mcg',   'Eggplant, raw'),
             (3,        '0',  'Teff, uncooked'),
             (4, '0.00 mcg', 'Sherbet, orange')],
            names=[None, 'folic_acid', 'name']),
 Index(['name', 'serving_size', 'calories', 'total_fat', 'saturated_fat',
        'cholesterol', 'sodium', 'choline', 'folate', 'folic_acid', 'niacin',
        'pantothenic_acid', 'riboflavin', 'thiamin', 'vitamin_a',
        'vitamin_a_rae', 'carotene_alpha', 'carotene_beta',
        'cryptoxanthin_beta', 'lutein_zeaxanthin', 'lucopene', 'vitamin_b12',
        'vitamin_b6', 'vitamin_c', 'vitamin_d', 'vitamin_e', 'tocopherol_alpha',
        'vitamin_k', 'calcium', 'copper', 'irom', 'magnesium', 'manganese',
        'phosphorous', 'potassium', 'selenium', 'zink', 'protein', 'alanine',
        'arginine', 'aspartic_acid', 'cystine', 'glutamic_acid', 'glycine',
        'histidine', 'hydroxyproline',

Lastly, the `set_index` method has a `verify_integrity` parameter. When set to `True`, Pandas will check whether the newly-created Index contains all unique values. If it does not, the method will throw an error.

To illustrate, let's use the `value_counts()` method to determine whether any foods share the same calorie count. Sure enough they do.

In [50]:
nutrition.calories.value_counts()

884    78
47     45
56     43
0      39
63     38
       ..
593     1
657     1
665     1
673     1
727     1
Name: calories, Length: 671, dtype: int64

Now let's try to set *calories* as the index.

In [51]:
# nutrition.set_index('calories', verify_integrity=True)

Although Pandas allows duplicates in the index, the instructor recommends *against* utilizing this feature. More often than not, it will lead to poor performance and errors down the road. For example, accessing observations by index label is no longer as useful.

## Extracting from DataFrames by Label

Many of the same methods that we explored for Series also apply to DataFrames! Even better, since DataFrames have two dimensions, we have even more power and options to slice and extract, since we can do so for both rows and columns.

Let's start with the `loc` attribute, which we saw many times in Series.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html

Let's first start by setting the `name` column as the index.


In [52]:
nutrition.head()

name serving_size  calories  ... caffeine theobromine    water
0       Cornstarch        100 g       381  ...  0.00 mg     0.00 mg   8.32 g
1     Nuts, pecans        100 g       691  ...  0.00 mg     0.00 mg   3.52 g
2    Eggplant, raw        100 g        25  ...  0.00 mg     0.00 mg  92.30 g
3   Teff, uncooked        100 g       367  ...        0           0   8.82 g
4  Sherbet, orange        100 g       144  ...  0.00 mg     0.00 mg  66.10 g

[5 rows x 76 columns]

In [53]:
nutrition.set_index('name', inplace=True)


In [54]:
nutrition.head()

serving_size  calories total_fat  ... caffeine theobromine    water
name                                              ...                              
Cornstarch             100 g       381      0.1g  ...  0.00 mg     0.00 mg   8.32 g
Nuts, pecans           100 g       691       72g  ...  0.00 mg     0.00 mg   3.52 g
Eggplant, raw          100 g        25      0.2g  ...  0.00 mg     0.00 mg  92.30 g
Teff, uncooked         100 g       367      2.4g  ...        0           0   8.82 g
Sherbet, orange        100 g       144        2g  ...  0.00 mg     0.00 mg  66.10 g

[5 rows x 75 columns]

Now use `loc` to access a particular entry by index label (i.e. column name).

In [55]:
nutrition.loc['Eggplant, raw']

serving_size       100 g
calories              25
total_fat           0.2g
saturated_fat        NaN
cholesterol            0
                  ...   
alcohol            0.0 g
ash               0.66 g
caffeine         0.00 mg
theobromine      0.00 mg
water            92.30 g
Name: Eggplant, raw, Length: 75, dtype: object

Note that what we get back is actually a Pandas Series, with the indices as the column names and the values as the value of those columns.

In [56]:
type(nutrition.loc['Eggplant, raw'])

pandas.core.series.Series

What if we want a **specific** entry using `loc`? That can be done by simply chaining on another set of square brackets that captures the index label (column name) of the property you want!

In [57]:
nutrition.loc['Eggplant, raw']['calories']

25

Alternatively, the `loc` attribute also allows you to access a specific entry all in one go. It accepts two arguments within the square brackets, and these can be thought of as row and column

`.loc[name_of_row, name_of_column]`

In [58]:
nutrition.loc['Eggplant, raw', 'calories']

25

We can also get slices of DataFrames by getting a little creative with `loc`. In this case, we have provided a range of labels in two dimensions, and so Pandas will return a subset of the DataFrame spanning those dimensions.

In [59]:
nutrition.loc['Eggplant, raw':'Sherbet, orange', 'calories':'cholesterol']

calories total_fat saturated_fat cholesterol
name                                                         
Eggplant, raw          25      0.2g           NaN           0
Teff, uncooked        367      2.4g          0.4g           0
Sherbet, orange       144        2g          1.2g         1mg

Another indexing technique is to pass in a **list** that specifies the rows and columns that you are interested in. The first list should contain the row labels that you want, while the second list should contain the column labels that you want.

This is VERY powerful. If you want just a few entries from your DataFrame and only specific columns, this is the way to do it.

In [60]:
nutrition.loc[
              ['Raspberries, raw'],
              ['protein', 'vitamin_b6']
]

protein vitamin_b6
name                               
Raspberries, raw  1.20 g   0.055 mg

In [61]:
nutrition.loc[
              ['Raspberries, raw','Blackberries, raw'],
              ['protein', 'vitamin_b6', 'water']
]

protein vitamin_b6    water
name                                         
Raspberries, raw   1.20 g   0.055 mg  85.75 g
Blackberries, raw  1.39 g   0.030 mg  88.15 g

## DataFrame Extraction by Position

We previously used `iloc` to extract from Series by position. We can do the same thing with DataFrames.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html

In [62]:
nutrition.head(10)

serving_size  calories  ... theobromine    water
name                                       ...                     
Cornstarch                100 g       381  ...     0.00 mg   8.32 g
Nuts, pecans              100 g       691  ...     0.00 mg   3.52 g
Eggplant, raw             100 g        25  ...     0.00 mg  92.30 g
Teff, uncooked            100 g       367  ...           0   8.82 g
Sherbet, orange           100 g       144  ...     0.00 mg  66.10 g
Cauliflower, raw          100 g        25  ...     0.00 mg  92.07 g
Taro leaves, raw          100 g        42  ...     0.00 mg  85.66 g
Lamb, raw, ground         100 g       282  ...     0.00 mg  59.47 g
Cheese, camembert         100 g       300  ...     0.00 mg  51.80 g
Vegetarian fillets        100 g       290  ...     0.00 mg  45.00 g

[10 rows x 75 columns]

Just like in Series, `iloc` indexing is zero-based for DataFrames. In the example below, we are indicating that we want the fourth row (indexed at 3) and all columns for that row.

In [63]:
nutrition.iloc[3]

serving_size      100 g
calories            367
total_fat          2.4g
saturated_fat      0.4g
cholesterol           0
                  ...  
alcohol               0
ash              2.37 g
caffeine              0
theobromine           0
water            8.82 g
Name: Teff, uncooked, Length: 75, dtype: object

We can grab multiple rows as well! To do this, we can pass in lists of row and column locations that we want. You can pass in either rows or columns, or both.

In [64]:
nutrition.iloc[
               [4,6,9]
]

serving_size  calories  ... theobromine    water
name                                       ...                     
Sherbet, orange           100 g       144  ...     0.00 mg  66.10 g
Taro leaves, raw          100 g        42  ...     0.00 mg  85.66 g
Vegetarian fillets        100 g       290  ...     0.00 mg  45.00 g

[3 rows x 75 columns]

Example of grabbing just one column. But remember that we can't use labels for `iloc`, so we need to know which integer position the column is at.

In [65]:
nutrition.iloc[
               [4,6,9],
               [2,3,4]
]

total_fat saturated_fat cholesterol
name                                                  
Sherbet, orange           2g          1.2g         1mg
Taro leaves, raw        0.7g          0.2g           0
Vegetarian fillets       18g          2.8g           0

Alternative syntax:

In [66]:
nutrition.iloc[
               [4,6,9],
               2:5
]

total_fat saturated_fat cholesterol
name                                                  
Sherbet, orange           2g          1.2g         1mg
Taro leaves, raw        0.7g          0.2g           0
Vegetarian fillets       18g          2.8g           0

Both the `loc` and `iloc` methods support boolean masks as well. Remember that when we used this for Series, we had to provide a Series of booleans that was exactly equal in size to the Series we are filtering.

Now that we're working with two-dimensional DataFrames, we need to provide *2 lists* of booleans. The first must be the same length as the number of rows, while the second must be the same length as the number of columns.

In this example, we'll create two lists of booleans such that we only grab the even rows and columns. As such, our new DataFrame will be about 1/4 the size of our old one.

In [67]:
evens_nutrition = nutrition.iloc[
               [True if i%2 == 0 else False for i in range(8789)],
               [True if i%2 == 0 else False for i in range(75)]
]
evens_nutrition

serving_size  ...    water
name                                                             ...         
Cornstarch                                                100 g  ...   8.32 g
Eggplant, raw                                             100 g  ...  92.30 g
Sherbet, orange                                           100 g  ...  66.10 g
Taro leaves, raw                                          100 g  ...  85.66 g
Cheese, camembert                                         100 g  ...  51.80 g
...                                                         ...  ...      ...
Beef, raw, select, trimmed to 1/8" fat, separab...        100 g  ...  70.89 g
Oil, uses similar to 95 degree hard butter, con...        100 g  ...   0.05 g
Beef, raw, all grades, trimmed to 0" fat, separ...        100 g  ...  72.51 g
Lamb, raw, separable lean and fat, composite of...        100 g  ...  59.80 g
Beef, raw, all grades, trimmed to 0" fat, separ...        100 g  ...  73.43 g

[4395 rows x 38 columns]

In [68]:
evens_nutrition.shape

(4395, 38)

What if we want to extract just one ONE SPECIFIC VALUE using `iloc`? Easy enough, just identify the coordinate for that value, then pass the coordinate to `iloc`.
* Pandas also has some built-in attributes that perform single-value extraction much more quickly than `iloc`. We'll cover these in the next lecture.

In [69]:
nutrition.iloc[9,1]

290

## Single Value Access With `.at` and `.iat`

Pandas offers two highly efficient alternative indexing attributes to extract single values, called `.at` and `.iat`

`.at` is an alternative to `.loc`, and similarly uses labeled-based search. It only works for accessing single values. You cannot pass in lists of indices or column labels.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html

In [70]:
nutrition.at['Nuts, pecans', 'calories']

691

In [71]:
nutrition.loc['Nuts, pecans', 'calories']

691

Similarly, we use `.iat` as an alternative to `.iloc`.

In [72]:
nutrition.iat[1,1]

691

Why use `.at` or `.iat` instead of `.loc` and `.iloc`, which can do the exact same thing? The answer is that `.at` and `.iat` are very specialized, single-purpose, and faster. By contrast, `.loc` and `.iloc` have a lot of computational "overhead" due to the fact that they have to interpret the input syntax and respond accordingly.

Observe this comparison using the `timeit` module
* https://docs.python.org/3/library/timeit.html

In [73]:
%timeit nutrition.loc['Nuts, pecans', 'calories']

The slowest run took 10.58 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 7.12 µs per loop


In [74]:
%timeit nutrition.at['Nuts, pecans', 'calories']

The slowest run took 20.66 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 3.81 µs per loop


Obviously these differences for a small dataset are miniscule and insignificant - `.at` is only about 4 microseconds faster than `.loc`. But these differences can add up when analyzing very large datasets!

## BONUS - The `get_loc()` Method

When working with position-based extraction approaches, you may find yourself counting columns and rows to find the exact integer location. Similar for slicing. This is especially the case when we combine labels with integer positions to find what we want.

For example, say we're interested in how much Vitamin K the third food in the dataset has. We know the column is `vitamin_k` and the index position is `3`. However, all of the methods we have learned so far will take one or the other, not both!

We can either use a label-based finder first, then use positional finders, or vice versa.


### Approach #1: Getting integer label from integer position
Start by isolating the index, and obtaining the label that corresponds to the third index position (the third row)

In [75]:
index_label = nutrition.index[2]

Next, assign the column label. In this case we want `vitamin_k`

In [76]:
column_label = 'vitamin_k'

Finally, use `.loc` with both of your labels to get the datapoint that you want.

In [77]:
nutrition.loc[index_label, column_label]

'3.5 mcg'

Alternatively, since we're using a single-value axis, we can also use the `.at` attribute. It is computational more efficient

In [78]:
nutrition.at[index_label, column_label]

'3.5 mcg'

### Approach #2: Determine the integer location from the label
Start by isolating the columns and obtaining the integer position corresponding with the column you want, in this case `vitamin_k`. 

We can do this using the `get_loc()` method, which returns the integer location, slice, or boolean mask of the requested label.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.get_loc.html

In [79]:
column_loc = nutrition.columns.get_loc('vitamin_k')
column_loc

26

Next, assign the index location. This is equal to `2` since we want the third entry in the DataFrame

In [80]:
index_loc = 2

Finally, use either `iloc` or `iat` to get the value!

In [81]:
print(nutrition.iloc[index_loc, column_loc])
print(nutrition.iat[index_loc, column_loc])

3.5 mcg
3.5 mcg


## More Cleanup: Going Numeric

In its current structure, our DataFrame is not conducive to numerical analysis. Why? Because although the values in these columns may initially look like numbers, most if not all all of these columns have values that are *strings*.

In [82]:
nutrition.head(5)

serving_size  calories total_fat  ... caffeine theobromine    water
name                                              ...                              
Cornstarch             100 g       381      0.1g  ...  0.00 mg     0.00 mg   8.32 g
Nuts, pecans           100 g       691       72g  ...  0.00 mg     0.00 mg   3.52 g
Eggplant, raw          100 g        25      0.2g  ...  0.00 mg     0.00 mg  92.30 g
Teff, uncooked         100 g       367      2.4g  ...        0           0   8.82 g
Sherbet, orange        100 g       144        2g  ...  0.00 mg     0.00 mg  66.10 g

[5 rows x 75 columns]

If you tried to add values together, you'll get a concatenated string, just as you would with string concatenation in Python.

In [83]:
nutrition.total_fat.sum()

'0.1g72g0.2g2.4g2g0.3g0.7g23g24g18g0g0.4g0.1g7.2g15g4.5g16g0.4g0.9g9.2g0.5g1.7g5.9g33g15g0g0g3g0.4g68g1.5g0.2g11g0.8g16g0.2g0.5g50g0.3g0.6g3g16g3.1g11g2.4g0g0.3g0.5g99g0.1g5.3g5.3g0.4g5.3g0.2g0.3g6.4g22g1g14g10g17g14g4.3g22g36g0.7g100g27g2.8g5.8g14g0.7g1.5g11g34g0.2g0.9g4.6g0g0.4g19g0.2g10g28g6.7g6.7g0.3g0.3g44g37g14g50g0.9g0.7g0.1g0.7g8g10g19g9.2g0g0.2g8.6g1.3g14g0.4g0.2g8.2g5.2g3.3g2.1g0.1g80g0.5g1.4g0.4g3.3g0.2g0.7g0.1g9.4g0.3g0.6g29g0.2g1.4g0.2g0.4g1.2g1.8g0g0.9g0.2g1.4g0.8g1.5g9.8g0g13g16g19g7.4g0.2g5.2g29g0.3g9.9g22g14g15g4.1g0.5g3.5g15g20g32g1g81g0.2g1.6g0g17g22g7.1g7.4g0.2g8.7g1.4g34g0.3g6.3g30g8.1g0.2g0.1g3.7g0.6g0.3g3.9g0.3g1.2g29g14g26g1.1g2g13g0g9g22g3.7g100g0.1g2.1g2.1g0.3g0.2g6.8g8.1g8.3g2.4g0.2g23g6.7g1g0.3g0.2g0.1g12g17g0.1g100g3.4g0.2g6.7g1.5g22g0g1g25g34g3.6g100g7.3g9g1.6g0.4g0.3g18g0.4g9.5g1.4g11g3.1g1.5g0g2.4g3.6g0g15g1.2g6.6g0.7g4.2g15g0.1g0.5g0.2g0.1g3.5g3g0.1g0.4g0.1g0.3g100g31g2.8g9.7g2.3g11g1.7g0.1g1.7g0.5g4.7g0.5g0.2g8.6g18g25g0.5g0.5g26g7.7g0.1g0.1g100g1.1g16

In order to do any math on this, we need to convert the data into *numeric*. Using the `.info()` method, we can see that most of the datatypes are `object` instead of `int` or `float`
*  https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html

In [84]:
nutrition.info(verbose = False)

<class 'pandas.core.frame.DataFrame'>
Index: 8789 entries, Cornstarch to Beef, raw, all grades, trimmed to 0" fat, separable lean only, boneless, eye of round steak, round
Columns: 75 entries, serving_size to water
dtypes: int64(2), object(73)
memory usage: 5.4+ MB


## The `astype()` Method

This is a very useful built-in method called `astype()` that allows us to cast a given DataFrame or Series from one datatype to another.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.astype.html

Let's first explore this method in isolation. Start by creating a new small DataFrame with three columns. Notice that *age* and *weight* are numeric (int and float, respectively) while *height* is a string (or more accurately, an object).

In [85]:
df = pd.DataFrame({
    'age': [12, 13, 14, 16],
    'weight': [41.1, 34.5, 83.2, 90.1],
    'height': ['1.72', '1.74', '1.91', '1.54']
})
df

age  weight height
0   12    41.1   1.72
1   13    34.5   1.74
2   14    83.2   1.91
3   16    90.1   1.54

Confirm the datatypes using the `info()` method

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     4 non-null      int64  
 1   weight  4 non-null      float64
 2   height  4 non-null      object 
dtypes: float64(1), int64(1), object(1)
memory usage: 224.0+ bytes


We can now use the `astype()` method to convert all columns to one common datatype. In this case we'll use **float**

In [87]:
df = df.astype(float)
df

age  weight  height
0  12.0    41.1    1.72
1  13.0    34.5    1.74
2  14.0    83.2    1.91
3  16.0    90.1    1.54

To check whether it worked, use the `info()` method again.

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     4 non-null      float64
 1   weight  4 non-null      float64
 2   height  4 non-null      float64
dtypes: float64(3)
memory usage: 224.0 bytes


The `astype()` method is also quite flexible. We can cast datatypes based on specific columns by passing in a dictionary whose keys are column names and values are datatypes.
* This works with both Python types and Numpy dtypes

In [89]:
df = df.astype({'age' : int})

In [90]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     4 non-null      int64  
 1   weight  4 non-null      float64
 2   height  4 non-null      float64
dtypes: float64(2), int64(1)
memory usage: 224.0 bytes


Now let's try to solve our problem with the `nutrition` DataFrame. Start by bringing up the first few rows and all columns, and attempt to case them to **floats**

In [91]:
# nutrition.iloc[0:4, :].astype(float)

Notice how this does NOT work. The reason is that the values are strings that contain things other than numbers, namely letters. Pandas does not know how to convert "100 g" to "100". 

For that, we will need to use a different method and regular expressions in order to remove the letters from the values.

## DataFrame `replace()` + A Glimpse at Regex

We can't use `astype()` as a one-stop solution to convert our values the way we want to. Instead, we'll need to separate the numerical values from the letters they are attached to, and then apply `astype()` to the numerical portion.

For this, we can use the `replace()` method. This intuitive method replaces one thing with another.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.replace.html

Let's start with some basic practice with `replace()`. Let's obtain the first six rows of our **nutrition** DataFrame and the first column.

In [92]:
dfm = nutrition.iloc[:6,0:1]
dfm

serving_size
name                         
Cornstarch              100 g
Nuts, pecans            100 g
Eggplant, raw           100 g
Teff, uncooked          100 g
Sherbet, orange         100 g
Cauliflower, raw        100 g

The *serving_size* column is of `object` datatype because they are all strings

In [93]:
nutrition.iloc[:6,0:1].info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, Cornstarch to Cauliflower, raw
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   serving_size  6 non-null      object
dtypes: object(1)
memory usage: 96.0+ bytes


`replace()` works by passing an argument that contains what you want to replace to the `to_replace` parameter, and the value you want to replace it with to the `value` parameter.
* Note that you can also pass in `to_replace` and `value` parameters positionally, as they are the first two positional parameters for the method

In [94]:
dfm.replace(to_replace='100 g', value = 100)

serving_size
name                          
Cornstarch                 100
Nuts, pecans               100
Eggplant, raw              100
Teff, uncooked             100
Sherbet, orange            100
Cauliflower, raw           100

Let's check the datatype of this replaced column. Sure enough, we have converted these values to integers.

In [95]:
dfm.replace(to_replace='100 g', value = 100).info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, Cornstarch to Cauliflower, raw
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   serving_size  6 non-null      int64
dtypes: int64(1)
memory usage: 96.0+ bytes


But there's an issue: using `replace()` in this method is very targeted. That is, it depends on EXACT STRINGS. More useful would be to replace the *g* with nothing, leaving just the numerical portion. Let's try it:

In [96]:
dfm.replace('g', '')

serving_size
name                         
Cornstarch              100 g
Nuts, pecans            100 g
Eggplant, raw           100 g
Teff, uncooked          100 g
Sherbet, orange         100 g
Cauliflower, raw        100 g

This clearly did not work, the reason being that the method is looking for a value that exactly matches *g*. This does not exist in our DataFrame, as each instance of *g* is part of a larger string. Thus, the method just returns a copy of the same DataFrame.

Instead, we can use *regular expressions* to target and replace certain patterns of strings, not the exact strings themselves. 

The `replace()` method remains the same. But now we invoke the `regex` parameter and set it to `True`, telling the method that we want the `to_replace` and `value` parameters to be interpreted as regular expressions.

What regex should we use for this? Remember that we want to remove the space after the number as well as the g after the space and replace that selection with nothing. That can be identified using the `\sg` regular expression.

In [97]:
dfm.replace('\sg', '', regex = True)

serving_size
name                         
Cornstarch                100
Nuts, pecans              100
Eggplant, raw             100
Teff, uncooked            100
Sherbet, orange           100
Cauliflower, raw          100

Viola, the ` g` is gone! We are now ready to cast this entire simple DataFrame to `int` using `astype()`

In [98]:
dfm.replace('\sg', '', regex = True).astype(int)

serving_size
name                          
Cornstarch                 100
Nuts, pecans               100
Eggplant, raw              100
Teff, uncooked             100
Sherbet, orange            100
Cauliflower, raw           100

But are we ready to scale this sucker up to the entire **nutrition** DataFrame? Not really, because we have many patterns that need to be replaced. They include:
* g
* g (with no space)
* mg
* mcg

We would need to identify and replace ALL of these patterns! `replace()` can be used for this, but there are better ways that we will see later on. Also, do we REALLY want to get rid of the units entirely? What if we forget what the units were for each column? We probably want to keep the units somewhere, and one good place to sock them is into the column names!

## Part 1 of Replacing the Values - Collecting the Units

In this effort, we will extract the units from each column. Yes, it will feel tedious and challenging, but that's a good thing.

So how do we isolate the units? Let's start by removing all the numerical stuff. If we remove the numbers, only the units will remain.

One way to do this is the use the `replace()` method with a regex pattern that identifies the numeric portion of each string and removes it.

Let's grab a sample of our DataFrame to see what we're dealing with.

In [99]:
nutrition.sample(20, axis = 1).head()

leucine saturated_fatty_acids  ...   serine  glycine
name                                            ...                  
Cornstarch       0.036 g               0.009 g  ...  0.012 g  0.009 g
Nuts, pecans     0.598 g               6.180 g  ...  0.474 g  0.453 g
Eggplant, raw    0.064 g               0.034 g  ...  0.042 g  0.041 g
Teff, uncooked   1.068 g               0.449 g  ...  0.622 g  0.477 g
Sherbet, orange        0               1.160 g  ...        0        0

[5 rows x 20 columns]

We've got a hodgepodge of different values. Some have no letters. Some have *g*. Some have *mg*. Others have international units (*IU*). Some have spaces between the number and the unit. Others do not. 

Which regular expression will select ALL potential numeric characters? We can do this using **character sets** in regex, which allow you to specify a group of characters that you want to match. We can do this with the following regex:
`\s*[^a-z+A-Z+]`. This essentially says "find any string that starts with any number of letters A through Z (lowercase or capitalized) which may or may not be preceded by a whitespace. Then, replace anything that is NOT part of this selection with a blank". Put another way, "keep all letters, remove all numbers and replace with blank".

In [100]:
nutrition.replace('\s*[^a-z+A-Z+]', '', regex = True).head(5)

serving_size  calories total_fat  ... caffeine theobromine water
name                                              ...                           
Cornstarch                 g       381         g  ...       mg          mg     g
Nuts, pecans               g       691         g  ...       mg          mg     g
Eggplant, raw              g        25         g  ...       mg          mg     g
Teff, uncooked             g       367         g  ...                          g
Sherbet, orange            g       144         g  ...       mg          mg     g

[5 rows x 75 columns]

Note that the **calories** column is numeric and was not affected by the regex search. We can deal with it by casting the entire DataFrame as strings before we do the replacement, such that all columns will be impacted by the regex replacement. Remember, we are not doing anything with these numbers for the moment. We just want the units.

In [101]:
units = nutrition.astype(str).replace('\s*[^a-z+A-Z+]', '', regex = True)
units

serving_size  ... water
name                                                             ...      
Cornstarch                                                    g  ...     g
Nuts, pecans                                                  g  ...     g
Eggplant, raw                                                 g  ...     g
Teff, uncooked                                                g  ...     g
Sherbet, orange                                               g  ...     g
...                                                         ...  ...   ...
Beef, raw, all grades, trimmed to 0" fat, separ...            g  ...     g
Lamb, cooked, separable lean only, composite of...            g  ...     g
Lamb, raw, separable lean and fat, composite of...            g  ...     g
Beef, raw, all grades, trimmed to 0" fat, separ...            g  ...     g
Beef, raw, all grades, trimmed to 0" fat, separ...            g  ...     g

[8789 rows x 75 columns]

We've now removed all numerical values. But notice that we still have some gaps.
* Sometimes units are included, sometimes they're not
* There are some instances of `NaN`
* There are some columns that are totally blank

How do we deal with this? We can't just pick a random row to represent our units, as it may have a gap in one or more of the columns. 

One way do deal with this is to find the *mode unit* from each column, that is, the unit that appears most often in each column. There is a good chance that the mode unit represents the unit that belongs in that column. For example, let's check out the **saturated_fat** column.

In [102]:
units.saturated_fat.value_counts()

g      7199
nan    1590
Name: saturated_fat, dtype: int64

We see here that 1590 foods (rows) have `NaN` (probably representing foods with no saturated fat, while the remaining 7199 rows have `g` as unit. Thus, we conclude that the most common unit is `g`. 

How do we collect the mode unit for every column? Simple! Apply the `mode()` method to the DataFrame. This will give you a DataFrame reporting the mode unit for each column in the original DataFrame. Amazing!
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mode.html

In [103]:
units.mode()

serving_size calories total_fat saturated_fat  ... ash caffeine theobromine water
0            g                  g             g  ...   g       mg          mg     g

[1 rows x 75 columns]

Now let's reassign the `units` dataframe to point to the *unit mode* DataFrame. We've got our units and now can do some real damage with them.

We'll continue this exercise in the next section.

In [104]:
units = units.mode()
units

serving_size calories total_fat saturated_fat  ... ash caffeine theobromine water
0            g                  g             g  ...   g       mg          mg     g

[1 rows x 75 columns]

## The `rename()` Method

`rename()` is one of the most useful methods in Pandas. It allows us to alter the axes labels of either our index or our columns. This will be critical for adding the units to the columns in the `nutrition` DataFrame
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

Let's start with the simple `df` DataFrame from earlier.

In [105]:
df

age  weight  height
0   12    41.1    1.72
1   13    34.5    1.74
2   14    83.2    1.91
3   16    90.1    1.54

If we want to rename our indices, we can pass in to the `index` paramater a dictionary with the current index labels as the keys and the new labels as the values. 
* Keep in mind that like many other DataFrame methods, by default `rename()` creates a copy of the original DataFrame. To perform the renaming in place, use the familiar `inplace` parameter.

In [106]:
df.rename(index={0: 'Pikachu', 1: 'Andy'})

age  weight  height
Pikachu   12    41.1    1.72
Andy      13    34.5    1.74
2         14    83.2    1.91
3         16    90.1    1.54

How about renaming the column labels? Easy, just pass the dictionary into the `columns` parameter!

In [107]:
df.rename(columns = {'weight':'Weight (kg)'})

age  Weight (kg)  height
0   12         41.1    1.72
1   13         34.5    1.74
2   14         83.2    1.91
3   16         90.1    1.54

Can you perform renaming on both axes at the same time? You bet!

In [108]:
df.rename(index={0: 'Pikachu', 1: 'Andy'}, columns = {'weight':'Weight (kg)'})

age  Weight (kg)  height
Pikachu   12         41.1    1.72
Andy      13         34.5    1.74
2         14         83.2    1.91
3         16         90.1    1.54

### An alternative approach to renaming - the `mapper` parameter

Using the `mapper` parameter within rename is an alternative to using the `index` and `columns` parameters. To do this:
1. Pass a dictionary into `mapper` that describes the labels you want to replace, and what you want to replace them with (similar to `index` and `columns`)
2. Specify the axis that the mapper should apply to by using the `axis` paramter. `0` for "index" or `1` for "columns.

* Note that `mapper` cannot be used to rename both the rows and columns at the same time. In those cases we would be forced to use `columns` and `index`.

In [109]:
df.rename(mapper={'height':'Height (m)'}, axis=1)

age  weight  Height (m)
0   12    41.1        1.72
1   13    34.5        1.74
2   14    83.2        1.91
3   16    90.1        1.54

## DataFrame `dropna()`
We previously used the Series `dropna()` method to drop all `NaN` values for a Series. DataFrames also have a `dropna()` method that works the exact same way, but it comes packed with a few additional features. With DataFrames, we can specify the axis that we want to drop from.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html
 * This is yet another method that can utilize the `inplace` parameter.

Let's continue working with our small `df` DataFrame and insert an `NaN` value.

In [110]:
df.loc[2, 'weight'] = np.nan
df

age  weight  height
0   12    41.1    1.72
1   13    34.5    1.74
2   14     NaN    1.91
3   16    90.1    1.54

Let's also change an entire row to contain `NaN` values.

In [111]:
df.loc[1, :] = np.nan
df

age  weight  height
0  12.0    41.1    1.72
1   NaN     NaN     NaN
2  14.0     NaN    1.91
3  16.0    90.1    1.54

We now have a DataFrame with at least one `NaN` in each column and at least one `NaN` in two of its four rows. Let's see how `dropna()` works here.


In [112]:
df.dropna()

age  weight  height
0  12.0    41.1    1.72
3  16.0    90.1    1.54

As you can see, by default, dropna() will exclude any *row* that contains at least one `NaN` value. Additionally, the `how` parameter defaults to "any", meaning that if ANY row contains a `NaN` value, that row will be dropped. But we can customize this.
* We can select the axis on which the method operates (0/rows or 1/columns) with the `axis` parameter
* We can set the `how` parameter to "all", meaning the row/column will only get dropped if ALL columns in that row or all rows in that column contain `NaN`

In [113]:
df.dropna(how = 'any', axis = 0)

age  weight  height
0  12.0    41.1    1.72
3  16.0    90.1    1.54

If we set the axis to 0 (rows) and `how = 'all'`, only the row with all `NaN` values will be dropped.

In [114]:
df.dropna(how = 'all', axis = 0)

age  weight  height
0  12.0    41.1    1.72
2  14.0     NaN    1.91
3  16.0    90.1    1.54

Another parameter we can set is `thresh`. This lets us specify the minimum number of non-NA values that a record must contain in order to be kept in the DataFrame.

In [115]:
df.dropna(thresh=3, axis = 0)

age  weight  height
0  12.0    41.1    1.72
3  16.0    90.1    1.54

Let's play with the axis paremeter. What if we set `axis=1` for our small `df` DataFrame

The following code returns an empty DataFrame! Why? Because each column had at least one `NaN` value.

In [116]:
df.dropna(axis = 1)

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3]

Let's play with the `thresh` parameter for the column axis. In this code, all columns that do not have at least 3 non-NA values will be dropped.

In [117]:
df.dropna(axis = 1, thresh = 3, inplace = True)
df

age  height
0  12.0    1.72
1   NaN     NaN
2  14.0    1.91
3  16.0    1.54

## BONUS Lecture - `dropna()` with Subset

The `subset` parameter of `dropna()` allows you to specify labels along the non-dropping axis (the orthogonal axis) to consider.

Let's start by adding another column, **gender**, to our small DataFrame

In [118]:
df['gender'] = ['M', 'F', np.nan, 'F']
df

age  height gender
0  12.0    1.72      M
1   NaN     NaN      F
2  14.0    1.91    NaN
3  16.0    1.54      F

If we run `dropna()`, we will see that rows 1 and 2 will be dropped since both of them contain at least one `NaN`

In [119]:
df.dropna()

age  height gender
0  12.0    1.72      M
3  16.0    1.54      F

Using `subset`, we can restrict `dropna()` to only the rows and columns that interest us. In this example, we are telling the method to drop any row that has a `NaN` in the **gender** column. The method will not consider any other columns when performing this.

Thus, only the one row that has `NaN` in the **gender** column (row 2) will be dropped.

In [120]:
df.dropna(axis = 0, how = 'any', subset = ['gender'])

age  height gender
0  12.0    1.72      M
1   NaN     NaN      F
3  16.0    1.54      F

Trying again with **age** column

In [121]:
df.dropna(axis = 0, how = 'any', subset = ['age'])

age  height gender
0  12.0    1.72      M
2  14.0    1.91    NaN
3  16.0    1.54      F

This also works with the column axis. Here, we specify a list of index labels that we want to scrutinize for the drops. Let's try it.



In [122]:
df

age  height gender
0  12.0    1.72      M
1   NaN     NaN      F
2  14.0    1.91    NaN
3  16.0    1.54      F

Below, we are saying to drop any column that has a `NaN` value in either rows with index label 1 or 2. In this case, all columns have an `NaN` value in at least one of those rows, and so ALL columns will be dropped. 

In [123]:
df.dropna(axis = 1, how = 'any', subset = [1, 2])

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3]

In this example, we are saying to drop any column that has a `NaN` value in either rows with index label 0 or 2. Here, the **gender** column will be dropped because it has an `NaN` value in row 2. Row 0 has no `NaN` values, so it will not trigger any column drops.

In [124]:
df.dropna(axis = 1, how = 'any', subset = [0, 2])

age  height
0  12.0    1.72
1   NaN     NaN
2  14.0    1.91
3  16.0    1.54

## Part 2 of Replacing the Values: Merging Units with Column Names

Remember that our goal is to transform all of our data into numeric while also preserving the units by moving that information to the column names.

Recall our `units` DataFrame

In [125]:
units

serving_size calories total_fat saturated_fat  ... ash caffeine theobromine water
0            g                  g             g  ...   g       mg          mg     g

[1 rows x 75 columns]

We will now apply the `rename()` method. But first, we need to figure out how to iterate over the DataFrame so that we can do something with the units.

Using a `for loop`, we can grab all of the column labels from the `units` DataFrame.

In [126]:
for k in units:
  print(k)

serving_size
calories
total_fat
saturated_fat
cholesterol
sodium
choline
folate
folic_acid
niacin
pantothenic_acid
riboflavin
thiamin
vitamin_a
vitamin_a_rae
carotene_alpha
carotene_beta
cryptoxanthin_beta
lutein_zeaxanthin
lucopene
vitamin_b12
vitamin_b6
vitamin_c
vitamin_d
vitamin_e
tocopherol_alpha
vitamin_k
calcium
copper
irom
magnesium
manganese
phosphorous
potassium
selenium
zink
protein
alanine
arginine
aspartic_acid
cystine
glutamic_acid
glycine
histidine
hydroxyproline
isoleucine
leucine
lysine
methionine
phenylalanine
proline
serine
threonine
tryptophan
tyrosine
valine
carbohydrate
fiber
sugars
fructose
galactose
glucose
lactose
maltose
sucrose
fat
saturated_fatty_acids
monounsaturated_fatty_acids
polyunsaturated_fatty_acids
fatty_acids_total_trans
alcohol
ash
caffeine
theobromine
water


Furthermore, we can use square bracketing to get a Series for each column. We are selecting columns from the DataFrame, returning a Series.

In [127]:
for k in units:
  print(k, units[k])

serving_size 0    g
Name: serving_size, dtype: object
calories 0    
Name: calories, dtype: object
total_fat 0    g
Name: total_fat, dtype: object
saturated_fat 0    g
Name: saturated_fat, dtype: object
cholesterol 0    mg
Name: cholesterol, dtype: object
sodium 0    mg
Name: sodium, dtype: object
choline 0    mg
Name: choline, dtype: object
folate 0    mcg
Name: folate, dtype: object
folic_acid 0    mcg
Name: folic_acid, dtype: object
niacin 0    mg
Name: niacin, dtype: object
pantothenic_acid 0    mg
Name: pantothenic_acid, dtype: object
riboflavin 0    mg
Name: riboflavin, dtype: object
thiamin 0    mg
Name: thiamin, dtype: object
vitamin_a 0    IU
Name: vitamin_a, dtype: object
vitamin_a_rae 0    mcg
Name: vitamin_a_rae, dtype: object
carotene_alpha 0    mcg
Name: carotene_alpha, dtype: object
carotene_beta 0    mcg
Name: carotene_beta, dtype: object
cryptoxanthin_beta 0    mcg
Name: cryptoxanthin_beta, dtype: object
lutein_zeaxanthin 0    mcg
Name: lutein_zeaxanthin, dtype: object

Instead of the entire Series object, we just want the unit contained within the series. Where is that unit located? Let's look at just one of these Series to find out.

In [128]:
for item in units['total_fat']:
  print(item)

g


As we should have known, there is only one item in each column, indexed at the 0 position. Makes sense, since there was only one row in the original DataFrame! Let's grab the unit from each column.

In [129]:
for k in units:
  print(k, units[k].at[0])

serving_size g
calories 
total_fat g
saturated_fat g
cholesterol mg
sodium mg
choline mg
folate mcg
folic_acid mcg
niacin mg
pantothenic_acid mg
riboflavin mg
thiamin mg
vitamin_a IU
vitamin_a_rae mcg
carotene_alpha mcg
carotene_beta mcg
cryptoxanthin_beta mcg
lutein_zeaxanthin mcg
lucopene 
vitamin_b12 mcg
vitamin_b6 mg
vitamin_c mg
vitamin_d IU
vitamin_e mg
tocopherol_alpha mg
vitamin_k mcg
calcium mg
copper mg
irom mg
magnesium mg
manganese mg
phosphorous mg
potassium mg
selenium mcg
zink mg
protein g
alanine g
arginine g
aspartic_acid g
cystine g
glutamic_acid g
glycine g
histidine g
hydroxyproline 
isoleucine g
leucine g
lysine g
methionine g
phenylalanine g
proline g
serine g
threonine g
tryptophan g
tyrosine g
valine g
carbohydrate g
fiber g
sugars g
fructose 
galactose 
glucose 
lactose 
maltose 
sucrose 
fat g
saturated_fatty_acids g
monounsaturated_fatty_acids g
polyunsaturated_fatty_acids g
fatty_acids_total_trans mg
alcohol g
ash g
caffeine mg
theobromine mg
water g


Great! Let's now remove the column labels to see the units in isolation. When doing this, we see lots of gaps - some characteristics do not have any units.

In [130]:
for k in units:
  print(units[k].at[0])

g

g
g
mg
mg
mg
mcg
mcg
mg
mg
mg
mg
IU
mcg
mcg
mcg
mcg
mcg

mcg
mg
mg
IU
mg
mg
mcg
mg
mg
mg
mg
mg
mg
mg
mcg
mg
g
g
g
g
g
g
g
g

g
g
g
g
g
g
g
g
g
g
g
g
g
g






g
g
g
g
mg
g
g
mg
mg
g


Column labels describing food characteristics that have no units do not need to be modified. That is, we do not need to edit these labels to include units, as they do not have units. We can remove these labels from the `units` DataFrame.

To remove irrelevant columns, we can use the `replace()` method to replace the blanks with `NaN`. Then we will chain on the `dropna()` method to remove columns with `NaN`.

In [131]:
units = units.replace('', np.nan).dropna(axis=1)
units

serving_size total_fat saturated_fat  ... caffeine theobromine water
0            g         g             g  ...       mg          mg     g

[1 rows x 66 columns]

In [132]:
for k in units:
  print(units[k].at[0])

g
g
g
mg
mg
mg
mcg
mcg
mg
mg
mg
mg
IU
mcg
mcg
mcg
mcg
mcg
mcg
mg
mg
IU
mg
mg
mcg
mg
mg
mg
mg
mg
mg
mg
mcg
mg
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
g
mg
g
g
mg
mg
g


Now that we have collected all of the units in a form that we can easily iterate over, the next step will be to create a mapper consisting of dictionary key:value pairs with old labels and the new. This mapper will then be used on the original DataFrame in order to replace the original column labels with the "unified" column labels.

Let's build our dictionary. Below is a conceptual example of the dictionary we are trying to build.

In [133]:
{
    'serving_size':'serving_size_g',
    'total_fat': 'total_fat_g'
}

{'serving_size': 'serving_size_g', 'total_fat': 'total_fat_g'}

Let's use **dictionary comprehension** to build this dictionary easily and quickly. Here, for each column in `units`, capture the column label `k` as well as the unit for that column

In [134]:
mapper = {k:units[k].at[0] for k in units}
mapper

{'alanine': 'g',
 'alcohol': 'g',
 'arginine': 'g',
 'ash': 'g',
 'aspartic_acid': 'g',
 'caffeine': 'mg',
 'calcium': 'mg',
 'carbohydrate': 'g',
 'carotene_alpha': 'mcg',
 'carotene_beta': 'mcg',
 'cholesterol': 'mg',
 'choline': 'mg',
 'copper': 'mg',
 'cryptoxanthin_beta': 'mcg',
 'cystine': 'g',
 'fat': 'g',
 'fatty_acids_total_trans': 'mg',
 'fiber': 'g',
 'folate': 'mcg',
 'folic_acid': 'mcg',
 'glutamic_acid': 'g',
 'glycine': 'g',
 'histidine': 'g',
 'irom': 'mg',
 'isoleucine': 'g',
 'leucine': 'g',
 'lutein_zeaxanthin': 'mcg',
 'lysine': 'g',
 'magnesium': 'mg',
 'manganese': 'mg',
 'methionine': 'g',
 'monounsaturated_fatty_acids': 'g',
 'niacin': 'mg',
 'pantothenic_acid': 'mg',
 'phenylalanine': 'g',
 'phosphorous': 'mg',
 'polyunsaturated_fatty_acids': 'g',
 'potassium': 'mg',
 'proline': 'g',
 'protein': 'g',
 'riboflavin': 'mg',
 'saturated_fat': 'g',
 'saturated_fatty_acids': 'g',
 'selenium': 'mcg',
 'serine': 'g',
 'serving_size': 'g',
 'sodium': 'mg',
 'sugars': 'g

This is getting close. But we don't just want the units in isolation because we wouldn't have any idea of what the nutrition characteristic is. Instead, we want to *append* the units to the existing column names.

In [135]:
mapper = {k: k + "_" + units[k].at[0] for k in units}
mapper

{'alanine': 'alanine_g',
 'alcohol': 'alcohol_g',
 'arginine': 'arginine_g',
 'ash': 'ash_g',
 'aspartic_acid': 'aspartic_acid_g',
 'caffeine': 'caffeine_mg',
 'calcium': 'calcium_mg',
 'carbohydrate': 'carbohydrate_g',
 'carotene_alpha': 'carotene_alpha_mcg',
 'carotene_beta': 'carotene_beta_mcg',
 'cholesterol': 'cholesterol_mg',
 'choline': 'choline_mg',
 'copper': 'copper_mg',
 'cryptoxanthin_beta': 'cryptoxanthin_beta_mcg',
 'cystine': 'cystine_g',
 'fat': 'fat_g',
 'fatty_acids_total_trans': 'fatty_acids_total_trans_mg',
 'fiber': 'fiber_g',
 'folate': 'folate_mcg',
 'folic_acid': 'folic_acid_mcg',
 'glutamic_acid': 'glutamic_acid_g',
 'glycine': 'glycine_g',
 'histidine': 'histidine_g',
 'irom': 'irom_mg',
 'isoleucine': 'isoleucine_g',
 'leucine': 'leucine_g',
 'lutein_zeaxanthin': 'lutein_zeaxanthin_mcg',
 'lysine': 'lysine_g',
 'magnesium': 'magnesium_mg',
 'manganese': 'manganese_mg',
 'methionine': 'methionine_g',
 'monounsaturated_fatty_acids': 'monounsaturated_fatty_acids

Now for the grand almost-finale: let's use the `rename()` method to rename the columns of our `nutrition` DataFrame using our mapper dictionary!

In [136]:
nutrition.rename(columns=mapper, inplace = True)
nutrition

serving_size_g  ...  water_g
name                                                               ...         
Cornstarch                                                  100 g  ...   8.32 g
Nuts, pecans                                                100 g  ...   3.52 g
Eggplant, raw                                               100 g  ...  92.30 g
Teff, uncooked                                              100 g  ...   8.82 g
Sherbet, orange                                             100 g  ...  66.10 g
...                                                           ...  ...      ...
Beef, raw, all grades, trimmed to 0" fat, separ...          100 g  ...  72.51 g
Lamb, cooked, separable lean only, composite of...          100 g  ...  59.95 g
Lamb, raw, separable lean and fat, composite of...          100 g  ...  59.80 g
Beef, raw, all grades, trimmed to 0" fat, separ...          100 g  ...  73.43 g
Beef, raw, all grades, trimmed to 0" fat, separ...          100 g  ...  73.43 g

[8789 rows x 75 columns]

Look at that, so beautiful and elegant. The best part is that only the columns that have units are are impacted by this rename. Any columns that have no units are left alone, and that's exactly what we want.

## Part 3 of Replacing the Values: Removing Units from Values

In this last part, we will remove all of the units from the original DataFrame. Nothing new, just an exercise of the methods we've already used.

Our approach of choice will be using the `replace()` method using *regular expressions*.

Multiple regex patters are possible, but to identify any letter values we can use `[a-z+A-Z+]`. Then us the `replace()` method to state that any pattern matching the regex expression by replaced with a blank, leaving only numbers behind.

In [137]:
nutrition.replace('[a-z+A-Z+]', '', regex = True, inplace=True)

In [138]:
nutrition.head()

serving_size_g  calories  ... theobromine_mg water_g
name                                      ...                       
Cornstarch                100        381  ...          0.00    8.32 
Nuts, pecans              100        691  ...          0.00    3.52 
Eggplant, raw             100         25  ...          0.00   92.30 
Teff, uncooked            100        367  ...              0   8.82 
Sherbet, orange           100        144  ...          0.00   66.10 

[5 rows x 75 columns]

We're note quite done yet though. If we take a look at the datatypes, we'll see that 73 out of 74 of our columns are non-numeric, but rather are of dtype "object".

In [139]:
nutrition.dtypes.value_counts()

object    73
int64      2
dtype: int64

 To fix this, we will convert all values with numbers to *floats* using the `.astype()` method.

In [140]:
nutrition = nutrition.astype(float)
nutrition.head()

serving_size_g  calories  ...  theobromine_mg  water_g
name                                       ...                         
Cornstarch                100.0     381.0  ...             0.0     8.32
Nuts, pecans              100.0     691.0  ...             0.0     3.52
Eggplant, raw             100.0      25.0  ...             0.0    92.30
Teff, uncooked            100.0     367.0  ...             0.0     8.82
Sherbet, orange           100.0     144.0  ...             0.0    66.10

[5 rows x 75 columns]

Problem solved! Now every column is a float.

In [141]:
nutrition.dtypes.value_counts()

float64    75
dtype: int64

Let's try some basic math to demonstrate.

In [142]:
nutrition.total_fat_g.sum()

92784.20000000001

In [143]:
nutrition.info(verbose = False)

<class 'pandas.core.frame.DataFrame'>
Index: 8789 entries, Cornstarch to Beef, raw, all grades, trimmed to 0" fat, separable lean only, boneless, eye of round steak, round
Columns: 75 entries, serving_size_g to water_g
dtypes: float64(75)
memory usage: 5.4+ MB


## Filtering DataFrames in 2D with `filter()`

When you have very large datasets, it is nice to be able to slice and dice just the index labels or column labels that you are interested in. This can be done with the `filter()` method. It is similar to the Series filter method, but now we can identify both index and columns.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.filter.html

Suppose we want to scrutinize octopus, since we'be been told it has a lot of cholesterol. Let's search the axis labels for "Octopus"

In [145]:
nutrition.filter(like = 'Octopus', axis=0)

serving_size_g  calories  ...  theobromine_mg  water_g
name                                               ...                         
Octopus (Alaska Native)           100.0      56.0  ...             0.0     84.0

[1 rows x 75 columns]

Remember that the `like` parameter is case-literal. If we change it to lower-case, we'll get a slightly different result.

In [146]:
nutrition.filter(like = 'octopus', axis=0)

serving_size_g  ...  water_g
name                                                           ...         
Mollusks, raw, common, octopus                          100.0  ...    80.25
Mollusks, moist heat, cooked, common, octopus           100.0  ...    60.50

[2 rows x 75 columns]

What if we want to capture both? Luckily, `filter()` has a `regex` parameter that we can use. We can specify that we want to look for the word "octopus, both lower-case and upper-case"

In [149]:
nutrition.filter(regex = '[oO]ctopus', axis = 0)

serving_size_g  ...  water_g
name                                                           ...         
Octopus (Alaska Native)                                 100.0  ...    84.00
Mollusks, raw, common, octopus                          100.0  ...    80.25
Mollusks, moist heat, cooked, common, octopus           100.0  ...    60.50

[3 rows x 75 columns]

We can also ask regex to simply ignore cases entirely (this is a bit more advanced regex) by activating case-insensitive mode.
* http://www.rexegg.com/regex-quickstart.html

In [150]:
nutrition.filter(regex = '(?i)octopus', axis = 0)

serving_size_g  ...  water_g
name                                                           ...         
Octopus (Alaska Native)                                 100.0  ...    84.00
Mollusks, raw, common, octopus                          100.0  ...    80.25
Mollusks, moist heat, cooked, common, octopus           100.0  ...    60.50

[3 rows x 75 columns]

Can we filter along both dimensions? After all, we care about the cholesterol in octopuses, and maybe serving size and calories, but not much else. Yes, we can do this 2D filtration!

We will chain on another `filter()` method and make use of the `items` parameter where we can pass in a list of things that we want to filter for.

In [154]:
nutrition.filter(regex = '(?i)octopus', axis = 0).filter(items = ['cholesterol_mg', 'serving_size_g', 'calories'], axis = 1)

cholesterol_mg  ...  calories
name                                                           ...          
Octopus (Alaska Native)                                  41.0  ...      56.0
Mollusks, raw, common, octopus                           48.0  ...      82.0
Mollusks, moist heat, cooked, common, octopus            96.0  ...     164.0

[3 rows x 3 columns]

This example was illustrative just to demonstrate the abilities of the `filter()` method. A better way to get the above result would be to chain `loc[]` on the row-filtered DataFrame and then select the columns by label. The end result is the same.

In [157]:
nutrition.filter(regex = '(?i)octopus', axis = 0).loc[:, ['cholesterol_mg', 'serving_size_g', 'calories']]

cholesterol_mg  ...  calories
name                                                           ...          
Octopus (Alaska Native)                                  41.0  ...      56.0
Mollusks, raw, common, octopus                           48.0  ...      82.0
Mollusks, moist heat, cooked, common, octopus            96.0  ...     164.0

[3 rows x 3 columns]

## Sorting with DataFrames: `sort_values()`
Now that our DataFrame is fully numeric, we are free to apply all of those numerical methods that we couldn't before. 

To refresh our memory, let's start by sorting by **vitamin B12**. Remember that when we use the column label as an attribute to a DataFrame. we get a Series back that describes the column.

In [158]:
nutrition.vitamin_b12_mcg

name
Cornstarch                                                                                            0.00
Nuts, pecans                                                                                          0.00
Eggplant, raw                                                                                         0.00
Teff, uncooked                                                                                        0.00
Sherbet, orange                                                                                       0.13
                                                                                                      ... 
Beef, raw, all grades, trimmed to 0" fat, separable lean and fat, boneless, top round roast, round    1.64
Lamb, cooked, separable lean only, composite of trimmed retail cuts, frozen, imported, New Zealand    2.95
Lamb, raw, separable lean and fat, composite of trimmed retail cuts, frozen, imported, New Zealand    2.42
Beef, raw, all grades, trimmed t

Now let's sort by the vitamin B12 content values using `.sort_values()`

In [159]:
nutrition.vitamin_b12_mcg.sort_values()

name
Cornstarch                                                                           0.00
Apricots, stewed, sulfured, dehydrated (low-moisture)                                0.00
Cocoa, processed with alkali, unsweetened, dry powder                                0.00
Tomato products, with herbs and cheese, sauce, canned                                0.00
Mothbeans, without salt, boiled, cooked, mature seeds                                0.00
                                                                                    ...  
Veal, braised, cooked, liver, variety meats and by-products                         84.60
Lamb, pan-fried, cooked, liver, variety meats and by-products                       85.70
Lamb, raw, liver, variety meats and by-products                                     90.05
Beef, boiled, cooked, variety meats and by-products liver, imported, New Zealand    96.00
Mollusks, moist heat, cooked, mixed species, clam                                   98.89
Name:

`sort_values()` also works on DataFrames as a whole. However, since DataFrames are two-dimensional, we must provide the *axis* to sort by as well as the axis names to sort by. Otherwise the method will throw an error.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html

In our example, we will sort the rows (indexes) by caloric content. 

In [164]:
nutrition.sort_values(axis = 0, by = 'calories', ascending=False).head()

serving_size_g  calories  ...  theobromine_mg  water_g
name                                           ...                         
Fat, mutton tallow            100.0     902.0  ...             0.0      0.0
Fish oil, salmon              100.0     902.0  ...             0.0      0.0
Lard                          100.0     902.0  ...             0.0      0.0
Fat, beef tallow              100.0     902.0  ...             0.0      0.0
Fish oil, cod liver           100.0     902.0  ...             0.0      0.0

[5 rows x 75 columns]

Unsurprisingly, the highest-calorie foods have the most fat. Not very interesting. Let's sort by something else, like **sodium**, along with cholesterol. We can do this by passing in a list of column labels to sort by.

This constitutes **sorting by two columns at once**. 

In [165]:
nutrition.sort_values(axis = 0, by = ['cholesterol_mg', 'sodium_mg'], ascending=False).head()

serving_size_g  ...  water_g
name                                                                ...         
Veal, braised, cooked, brain, variety meats and...           100.0  ...    76.89
Beef, simmered, cooked, brain, variety meats an...           100.0  ...    74.86
Beef, raw, brain, variety meats and by-products              100.0  ...    76.29
Lamb, soaked and fried, cooked, brains, importe...           100.0  ...    73.11
Pork, braised, cooked, brain, variety meats and...           100.0  ...    75.88

[5 rows x 75 columns]

What's actually happening here when we sort by multiple columns? By default, the directioanality chosen by the `ascending` parameter applies to both columns. In that case, the first column "cholesterol_mg", is sorted first, and then second column "sodium_mg" is sorted. That means you may have a food with lower cholesterol but higher sodium, sicne cholesterol is first in the sort priority.

What if we wanted to sort two or more columns in different directions? We can do that by passing in a list of booleans to the `ascending` parameter!

In [167]:
nutrition.sort_values(axis = 0, by = ['cholesterol_mg', 'sodium_mg'], ascending=[False, True]).head(5)

serving_size_g  ...  water_g
name                                                                ...         
Beef, simmered, cooked, brain, variety meats an...           100.0  ...    74.86
Veal, braised, cooked, brain, variety meats and...           100.0  ...    76.89
Beef, raw, brain, variety meats and by-products              100.0  ...    76.29
Lamb, soaked and fried, cooked, brains, importe...           100.0  ...    73.11
Pork, braised, cooked, brain, variety meats and...           100.0  ...    75.88

[5 rows x 75 columns]

Comparied to the unidirectional sorting, we see that the bi-directionary sorting has shifted *beef brain* to be the highest ranked, followed by *veal brain*. Beef brain has the same amount of cholesterol as veal brain, but less sodium. Cool!

## Using Series `between()` with DataFrames

The `between()` method gives us a convenient way to specify a range of values that we want. It is a **Series** method, so it must be applied to a 1-dimensional Series slice of our DataFrame
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.between.html

Let's play around with **calories**

In [169]:
nutrition.calories.head(10)

name
Cornstarch            381.0
Nuts, pecans          691.0
Eggplant, raw          25.0
Teff, uncooked        367.0
Sherbet, orange       144.0
Cauliflower, raw       25.0
Taro leaves, raw       42.0
Lamb, raw, ground     282.0
Cheese, camembert     300.0
Vegetarian fillets    290.0
Name: calories, dtype: float64

Let's say we're looking for a low-calorie food, say between 20 to 60 calories per 100g serving (remember that the `nutrition` DataFrame is standardized to 100g). We do this by selecting the `calories` attribute (thus getting a Series) and then specifying the range of calories we want.

In [170]:
nutrition.calories.between(20, 60)

name
Cornstarch                                                                                            False
Nuts, pecans                                                                                          False
Eggplant, raw                                                                                          True
Teff, uncooked                                                                                        False
Sherbet, orange                                                                                       False
                                                                                                      ...  
Beef, raw, all grades, trimmed to 0" fat, separable lean and fat, boneless, top round roast, round    False
Lamb, cooked, separable lean only, composite of trimmed retail cuts, frozen, imported, New Zealand    False
Lamb, raw, separable lean and fat, composite of trimmed retail cuts, frozen, imported, New Zealand    False
Beef, raw, all grades, 

What returns is a **Boolean series** that is the same length as the `calories` Series, which tells us whether each food is within the range of 20-60 calories. Now we can use this as a boolean mask to filter for low-calorie foods on the original DataFrame! Any food corresponding to a `True` in the boolean mask will be kept.

In [174]:
nutrition[nutrition.calories.between(20, 60)].head()

serving_size_g  calories  ...  theobromine_mg  water_g
name                                            ...                         
Eggplant, raw                  100.0      25.0  ...             0.0    92.30
Cauliflower, raw               100.0      25.0  ...             0.0    92.07
Taro leaves, raw               100.0      42.0  ...             0.0    85.66
PACE, Picante Sauce            100.0      25.0  ...             0.0    89.90
Mango nectar, canned           100.0      51.0  ...             0.0    86.63

[5 rows x 75 columns]

The returned DataFrame is over 1000 lines, corresponding to 1000 `True`s in the boolean mask. We can make this more manageable by using `sample()`.

In [175]:
nutrition[nutrition.calories.between(20, 60)].sample(4)

serving_size_g  ...  water_g
name                                                                ...         
Capers, canned                                               100.0  ...    83.85
Beverages, Strawberry Banana, V8 V-FUSION Juices             100.0  ...    87.90
Babyfood, orange and pineapple, juice                        100.0  ...    87.30
Plums, solids and liquids, water pack, purple, ...           100.0  ...    88.35

[4 rows x 75 columns]